### ISSUE TITLE
Database update error sql statments

### ISSUE THREAD

**Issue by contaoacademy** [codefog/contao-news_categories#156]
If I run the install-tool I get errors with this statements:

```
ALTER TABLE tl_content CHANGE news_filterPreserve news_filterPreserve TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_content CHANGE news_filterCategories news_filterCategories TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_module CHANGE news_resetCategories news_resetCategories TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_module CHANGE news_filterPreserve news_filterPreserve TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_module CHANGE news_relatedCategories news_relatedCategories TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_module CHANGE news_filterCategories news_filterCategories TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_module CHANGE news_customCategories news_customCategories TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_news_archive CHANGE limitCategories limitCategories TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_news_category CHANGE excludeInRelated excludeInRelated TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_news_category CHANGE hideInReader hideInReader TINYINT(1) DEFAULT '0' NOT NULL
ALTER TABLE tl_news_category CHANGE hideInList hideInList TINYINT(1) DEFAULT '0' NOT NULL
```

```
Error in Logfile:
[2019-01-11 16:30:56] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\DriverException: "An exception occurred while executing 'ALTER TABLE tl_news_category CHANGE hideInList hideInList TINYINT(1) DEFAULT '0' NOT NULL':  SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'hideInList' at row 1" at …/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 106 {"exception":"[object] (Doctrine\\DBAL\\Exception\\DriverException(code: 0): An exception occurred while executing 'ALTER TABLE tl_news_category CHANGE hideInList hideInList TINYINT(1) DEFAULT '0' NOT NULL':\n\nSQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'hideInList' at row 1 at …/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:106, Doctrine\\DBAL\\Driver\\PDOException(code: HY000): SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'hideInList' at row 1 at …/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:90, PDOException(code: HY000): SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'hideInList' at row 1 at …/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:88)"} []
```

**Comment by qzminski** [comment]
Is it a brand new setup of the extension, or are you trying to update it from Contao 3 to 4?

**Comment by contaoacademy** [comment]
This was an update from contao 3.5 to contao 4.6

Very interesting is, that I did it again and there was no problem.
The problem seems to be on my local MAMP system and on the live server I use mariadb.

Should we close the ticket or how can I help you to reproduce this error?

**Comment by qzminski** [comment]
Try changing the value of the mentioned columns to `0` manually (e.g. via phpMyAdmin) and then running install tool. The problem might be that your sql server is running in strict mode and fails on trying to update the column definition. The `tinyint` column type requires an integer value, whereas it currently is any empty string `''` – a leftover from old extension version for Contao 3.

**Comment by contaoacademy** [comment]
Thx. I executed this db-query to fix this problem.
I will close this ticket.

```
update tl_content set news_filterPreserve = '0' where news_filterPreserve = '';
update tl_content set news_filterCategories = '0' where news_filterCategories = '';
update tl_module set news_resetCategories = '0' where news_resetCategories = '';
update tl_module set news_filterPreserve = '0' where news_resetCategories = '';
update tl_module set news_relatedCategories = '0' where news_relatedCategories = '';
update tl_module set news_filterCategories = '0' where news_filterCategories = '';
update tl_module set news_customCategories = '0' where news_customCategories = '';
update tl_news_archive set limitCategories = '0' where limitCategories = '';
update tl_news_category set excludeInRelated = '0' where excludeInRelated = '';
update tl_news_category set hideInReader = '0' where hideInReader = '';
update tl_news_category set hideInList = '0' where hideInList = '';
```

**Comment by fritzmg** [comment]
I ran into this problem as well - but updating from version 2.8 of the extension to 3.x. I think it would be helpful if the extension provided an upgrade script executes those change.

**Comment by fritzmg** [comment]
Btw. there was a small error in line 4 of your database statements @contaoacademy 

Fixed statements:
```sql
UPDATE tl_content SET news_filterPreserve = '0' WHERE news_filterPreserve = '';
UPDATE tl_content SET news_filterCategories = '0' WHERE news_filterCategories = '';
UPDATE tl_module SET news_resetCategories = '0' WHERE news_resetCategories = '';
UPDATE tl_module SET news_filterPreserve = '0' WHERE news_filterPreserve = '';
UPDATE tl_module SET news_relatedCategories = '0' WHERE news_relatedCategories = '';
UPDATE tl_module SET news_filterCategories = '0' WHERE news_filterCategories = '';
UPDATE tl_module SET news_customCategories = '0' WHERE news_customCategories = '';
UPDATE tl_news_archive SET limitCategories = '0' WHERE limitCategories = '';
UPDATE tl_news_category SET excludeInRelated = '0' WHERE excludeInRelated = '';
UPDATE tl_news_category SET hideInReader = '0' WHERE hideInReader = '';
UPDATE tl_news_category SET hideInList = '0' WHERE hideInList = '';
```

**Comment by contaoacademy** [comment]
> Btw. there was a small error in line 4 of your database statements

Thx.

**Comment by fritzmg** [comment]
In Contao 4.9 you could implement a migration for this ;)

**Comment by qzminski** [comment]
Yeah now that the Contao 3 is dead I don't think it makes much sense. Also, there doesn't seem to be that many users who experience this issue + the fix is pretty easy.

**Comment by fritzmg** [comment]
Hm, in theory it should affect everyone who updates from version 2.x to 3.x of the extension, since the boolean datatypes aren't compatible with Contao's "char" boolean.

**Comment by fritzmg** [comment]
#### Reproduction

1. Install Contao 4.9.
2. Execute `composer require codefog/contao-news_categories:^2.0`.
3. Execute `vendor/bin/contao-console contao:migrate --no-interaction`.
4. Execute `composer require codefog/contao-news_categories:^3.0`.
5. Execute `vendor/bin/contao-console contao:migrate --no-interaction`.

The following errors will occur:

```
[ERROR] An exception occurred while executing 'ALTER TABLE tl_content CHANGE news_filterCategories            
         news_filterCategories TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format:
         1292 Truncated incorrect INTEGER value: '    '                                                        
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_content CHANGE news_filterPreserve              
         news_filterPreserve TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format: 1292
         Truncated incorrect INTEGER value: '    '                                                             
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_module CHANGE news_filterCategories             
         news_filterCategories TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format:
         1292 Truncated incorrect INTEGER value: '    '                                                        
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_module CHANGE news_relatedCategories            
         news_relatedCategories TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format:
         1292 Truncated incorrect INTEGER value: '    '                                                        
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_module CHANGE news_filterPreserve               
         news_filterPreserve TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format: 1292
         Truncated incorrect INTEGER value: '    '                                                             
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_module CHANGE news_resetCategories              
         news_resetCategories TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format:
         1292 Truncated incorrect INTEGER value: '    '                                                        
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_module CHANGE news_customCategories             
         news_customCategories TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format:
         1292 Truncated incorrect INTEGER value: '    '                                                        
                                                                                                               

                                                                                                               
 [ERROR] An exception occurred while executing 'ALTER TABLE tl_news_archive CHANGE limitCategories             
         limitCategories TINYINT(1) DEFAULT '0' NOT NULL':

SQLSTATE[22007]: Invalid datetime format: 1292
         Truncated incorrect INTEGER value: '    '
```

I can provide a PR, when I have the time :). I think we could integrate a Contao 4.9 Migration, which will simply not be executed in Contao 4.4, but I am not sure. Unless you also want a migration in Contao 4.4, then it could be done in a `sqlCompileCommands` hook (like I have done here: https://github.com/dklemmt/contao_dk_mmenu/blob/3.1.4/src/EventListener/SqlCompileCommandsListener.php)

**Comment by qzminski** [comment]
This problem appears when you upgrade from the previous major version to the next major version and by doing so compatibility issues can be expected. It's true that this could be covered by automated script but I don't thing it's worth the effort. If you would like to provide a 4.9 migration then I would be of course very happy to merge such PR.

**Comment by qzminski** [comment]
Thanks to @fritzmg this should be fixed as of 3.4.10. Thus, I am unpinning this issue.
